### PR TITLE
remove discovery_method field from network config [1/n]

### PR DESCRIPTION
**Pull request by aluon**
### Description
Set the default `discovery_methods` to `onchain` if the field is not set and added some validation around `discovery_methods`

I want to eventually remove the `discovery_method` field in favor of `discovery_methods` which supports multiple methods. Since unknown fields will break config loading, I'm going to remove all usages of `discovery_method` in the configs before removing the `discovery_method` field from the struct. I'm planning to land this over 3 PRs:
1. Set the default `discovery_methods` to `onchain` and remove usages of `discovery_method` from `aptos-core` (this PR)
2. Remove usages of `discovery_method` from the `internal-ops` repo ([link](https://github.com/aptos-labs/internal-ops/pull/2243))
3. Remove the `discovery_method` field from the struct in `aptos-core` (link)

### Test Plan

`cargo test -p aptos-node`

### CODE REVIEW

**Comment by aluon** [comment]
Current dependencies on/for this PR:
* `main`
  * **PR #11259** <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/11259?utm_source=stack-comment-icon" target="_blank">Graphite</a> 
    * **PR #11350** <a href="https://app.graphite.dev/github/pr/aptos-labs/aptos-core/11350?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/aptos-labs/aptos-core/11350?utm_source=stack-comment).

File: config/src/config/test_data/validator.yaml [hunk 1]
```diff
@@ -38,7 +38,6 @@ full_node_networks:
           private: "vfn"
 
 validator_network:
-    discovery_method: "onchain"
     listen_address: "/ip4/0.0.0.0/tcp/6180"
```
Comment (JoshLind) [inline config/src/config/test_data/validator.yaml:41]: Don't we need to prepend `listen_address` with `-`? Looks like it was removed with the line deletion, but I'm not sure if the YAML parser will be happy now 😄

File: docker/compose/aptos-node/validator.yaml [hunk 2]
```diff
@@ -22,7 +22,6 @@ execution:
   genesis_file_location: "/opt/aptos/genesis/genesis.blob"
 
 validator_network:
-  discovery_method: "onchain"
```
Comment (JoshLind) [inline docker/compose/aptos-node/validator.yaml:25]: Same here? Add `-`?

File: testsuite/pangu_lib/template_testnet_files/validator.yaml [hunk 3]
```diff
@@ -32,7 +32,6 @@ full_node_networks:
 
 
 validator_network:
-  discovery_method: "onchain"
```
Comment (JoshLind) [inline testsuite/pangu_lib/template_testnet_files/validator.yaml:35]: Same here? `-`?

File: config/src/config/network_config.rs [hunk 4]
```diff
@@ -215,18 +218,40 @@ impl NetworkConfig {
...
-            self.discovery_methods
-                .iter()
-                .filter(|method| &&DiscoveryMethod::None != method)
-                .collect()
+        if self.discovery_method.is_some() {
+            warn!("discovery_method is deprecated, use discovery_methods instead")
+        }
+
+        // Check for invalid combinations of discovery methods.
+        let using_none_with_others = self
```
Comment (JoshLind) [inline config/src/config/network_config.rs:228]: When we remove this function, we should eventually add this check to the config sanitizer 😄

**Review by JoshLind — APPROVED** [review]
Thanks @aluon! Looks reasonable to me 😄 

Note: if it were me, I'd just remove the config field in this PR altogether. That way, the compiler will catch any missing errors. Also, we won't have a release for a couple weeks, which should be more than enough time. But, your call 😄

File: config/src/config/test_data/validator.yaml (hunk 1 above)
Comment (aluon) [reply config/src/config/test_data/validator.yaml:41]: This tripped me up too. GitHub shows a `-` to indicate that the line was removed in my commit.

I double checked to make sure that I preserved the `-` if it was in the original file

**Comment by aluon** [comment]
Added a note about the removal of `discovery_method` to the release notes https://www.notion.so/aptoslabs/Devnet-Release-Notes-e8f228493e154fe0910ce7e6a4eb3ddc

**Review by davidiw — APPROVED** [review]

**Pull request closed without merging**